ResearchKit/Common/ORKTaskViewController.m

```diff
@@ -893,9 +893,18 @@ - (void)showViewController:(ORKStepViewController *)viewController goForward:(BO
         }
     }
     
-    if (step.identifier && ![_managedStepIdentifiers.lastObject isEqualToString:step.identifier]) {
-        [_managedStepIdentifiers addObject:step.identifier];
+    NSString *stepIdentifier = step.identifier;
+    if (stepIdentifier) {
+        NSUInteger indexOfStep = [_managedStepIdentifiers indexOfObject:stepIdentifier];
+        if (indexOfStep == NSNotFound) {
+            [_managedStepIdentifiers addObject:stepIdentifier];
+        } else {
+            // If the step being shown has been visited before, remove all the managed step identifiers visited after it
+            NSUInteger numberManagedOfStepIdentifiers = _managedStepIdentifiers.count;
+            [_managedStepIdentifiers removeObjectsInRange:NSMakeRange(indexOfStep + 1, numberManagedOfStepIdentifiers - (indexOfStep + 1))];
```

Comment:
in the first place, I don't think a loop should be allowed or in any type of tasks, because:

- If a step was showed to user more than once (not because use click `back` button), it is equivalent to two steps sharing a **same** identifier. Which is strictly forbidden in RK framework.

Current `NSArray` based `_managedStepIdentifiers` does not prevent the loop happening, but it keeps the actual history of the steps has been visited. It is more like a web browser.  

With the implementation of removing current step from `_managedStepIdentifiers` stack when user hit `back button`,  the stack won't keep very large managed identifiers.